### PR TITLE
HDSD-6 Separate Multiple Invoice PDFs

### PR DESCRIPTION
**Pull request by HoDacVu90**

### CODE REVIEW

File: hds_core/src/us/hds/ap/core/persistence/dao/DocumentDAO.java [hunk 1]
```diff
@@ -33,4 +36,35 @@ protected Class<Document> getType() {
...
+		return result;
+	}
+	
+	/**
+	 * find a {@link Document} by its extraction processing id
+	 *
+	 * @param processingId
+	 * @return
+	 */
+	public Document findByextractionProcessId(String processingId) {
```
Comment (ptanhaxon) [inline hds_core/src/us/hds/ap/core/persistence/dao/DocumentDAO.java:61]: findBy**E**tractionProcessId - the letter **E** should be caplitalize.

File: hds_idp/config/rest-clients.yaml [hunk 2]
```diff
@@ -22,7 +8,7 @@ RestClients:
     Properties:
       JSON.Deserialization.FAIL_ON_INVALID_SUBTYPE: 'false'
     OpenAPI:
-      SpecUrl: file:///C:/Users/hdvu/Downloads/Generic%20Document%20Splitting-openapi.json
+      SpecUrl: file:///C:/Users/hdvu/Documents/Projects/NatifAI/Generic%20Document%20Splitting-openapi.json
```
Comment (ptanhaxon) [inline hds_idp/config/rest-clients.yaml:11]: Why do we use the personal folder at this place? The file should be located in the process

File: hds_idp/config/rest-clients.yaml [hunk 3]
```diff
@@ -41,3 +27,27 @@ RestClients:
       SpecUrl: file:///C:/Users/hdvu/Documents/Projects/NatifAI/Doc%20Capturing-openapi.json
```
Comment (ptanhaxon) [inline hds_idp/config/rest-clients.yaml:27]: Same as above.

File: hds_idp/config/rest-clients.yaml [hunk 4]
```diff
@@ -41,3 +27,27 @@ RestClients:
...
       Namespace: ai.natif.api.client
       ResolveFully: true
+  IDP (Invoice Extraction API):
+    UUID: e4661358-fdad-4e37-84d1-e4c445b2abec
```
Comment (ptanhaxon) [inline hds_idp/config/rest-clients.yaml:31]: This should be the workflow ID?

File: hds_idp/config/variables.yaml [hunk 5]
```diff
@@ -10,4 +10,8 @@ Variables:
    hds:
       idp:
          apiKey: ''
-         wait_for: ''
\ No newline at end of file
+         invoiceWorkflowId: ''
+         extractionResponseIncludes: 'extractions'
+         splittingResponseIncludes: 'sub-pdfs'
+         wait_for: 120
+         scanDirectory: 'c:/ocr/'
```
Comment (ptanhaxon) [inline hds_idp/config/variables.yaml:17]: Is this valid? I wonder about  the `c:/`.

File: hds_idp/processes/InvoiceExtraction.p.json [hunk 6]
```diff
@@ -46,26 +46,27 @@
       "name" : "Invoice Extraction",
       "config" : {
         "headers" : {
-          "Accept" : "*/*",
+          "Accept" : "application/json",
           "Authorization" : "\"ApiKey \" + ivy.var.get(\"hds.idp.apiKey\")"
         },
-        "clientId" : "2d9937ad-dbc9-4c54-b334-dbc8eb7d7c63",
+        "clientId" : "6716cbcb-5f40-490a-a168-2e7dc5188cc3",
```
Comment (ptanhaxon) [inline hds_idp/processes/InvoiceExtraction.p.json:52]: The client ID should be taken from the variable,

File: hds_idp/processes/InvoiceExtraction.p.json [hunk 7]
```diff
@@ -46,26 +46,27 @@
...
         "method" : "POST",
         "queryParams" : {
-          "include" : "",
-          "wait_for" : ""
+          "include" : "us.hds.ap.idp.service.IDPService.getInvoiceExtractionInclude()"
         },
         "statusErrorCode" : "ivy:error:rest:client",
-        "bodyObjectType" : "ai.natif.api.client.BodyTProcessingInvoiceExtractionPost",
+        "bodyObjectType" : "ai.natif.api.client.BodyTProcessingEae93eaf991f496386ab11263de0a9d4Post",
```
Comment (ptanhaxon) [inline hds_idp/processes/InvoiceExtraction.p.json:58]: The class looks weird.

File: hds_idp/processes/InvoiceExtraction.p.json [hunk 8]
```diff
@@ -147,7 +158,123 @@
...
+      "id" : "f15",
+      "type" : "RestClientCall",
+      "name" : "Extraction Result",
+      "config" : {
+        "path" : "/processing/results/{processing_id}/extractions",
+        "headers" : {
+          "Accept" : "*/*",
+          "Authorization" : "\"ApiKey \" + ivy.var.get(\"hds.idp.apiKey\")"
+        },
+        "clientId" : "e4661358-fdad-4e37-84d1-e4c445b2abec",
```
Comment (ptanhaxon) [inline hds_idp/processes/InvoiceExtraction.p.json:192]: The client ID should be taken from variable.

File: hds_idp/processes/InvoiceExtraction.p.json [hunk 9]
```diff
@@ -147,7 +158,123 @@
...
+        "bodyRaw" : "<%=in.invoice%>",
+        "path" : "/documents/{documentId}/extractions",
+        "headers" : {
+          "Accept" : "*/*",
+          "Authorization" : "\"ApiKey \" + ivy.var.get(\"hds.idp.apiKey\")"
+        },
+        "bodyObjectMapping" : {
+          "param" : "in.invoice"
+        },
+        "clientId" : "6716cbcb-5f40-490a-a168-2e7dc5188cc3",
```
Comment (ptanhaxon) [inline hds_idp/processes/InvoiceExtraction.p.json:254]: The client ID should be taken from variable.

File: hds_idp/processes/InvoiceExtraction.p.json (hunk 8 above)
Comment (HoDacVu90) [reply hds_idp/processes/InvoiceExtraction.p.json:192]: It's the UUID of Ivy Rest client which has been called in this process

File: hds_idp/src/us/hds/ap/idp/service/IDPService.java [hunk 10]
```diff
@@ -1,22 +1,48 @@
...
-		return List.of("include=extractions","include=ocr","include=page-images","include=thumbnail","include=pdf");
+	public final static ObjectMapper objectMapper = new ObjectMapper();
+	
+	public static List<String> getInvoiceExtractionIncludes() {
+		String include = Ivy.var().get("hds.idp.extractionResponseIncludes");
+		return List.of(include);
 		
 	}
 	
+	public static String getInvoiceExtractionInclude() {
```
Comment (ptanhaxon) [inline hds_idp/src/us/hds/ap/idp/service/IDPService.java:33]: Do we use this function? I find that we have 2 functions with different returning.

File: hds_idp/processes/InvoiceExtraction.p.json (hunk 6 above)
Comment (HoDacVu90) [reply hds_idp/processes/InvoiceExtraction.p.json:52]: It's the UUID of Ivy Rest client which has been called in this process

File: hds_idp/src/us/hds/ap/idp/service/IDPService.java [hunk 11]
```diff
@@ -33,6 +59,10 @@ public static String getFileNameFromHeaderResponse(String str) {
 	}
 
 	public static String getSharingTokenUrl(UUID documentId, String token) {
+        return getSharingTokenUrl(documentId, token);
+	}
+	
+	public static String getSharingTokenUrl(String documentId, String token) {
         String url = StringUtils.EMPTY;
         String allowCookies = "&allowCookies=all";
```
Comment (ptanhaxon) [inline hds_idp/src/us/hds/ap/idp/service/IDPService.java:67]: Please define a constant.

File: hds_idp/config/variables.yaml (hunk 5 above)
Comment (HoDacVu90) [reply hds_idp/config/variables.yaml:17]: It working, but the var is used to testing by get file form disk c

File: hds_idp/config/rest-clients.yaml (hunk 4 above)
Comment (HoDacVu90) [reply hds_idp/config/rest-clients.yaml:31]: It's the UUID of Ivy Rest client

File: hds_idp/src/us/hds/ap/idp/service/IDPService.java [hunk 12]
```diff
@@ -53,4 +83,23 @@ public static String getTokenExpiredDate() {
 		return nextMonth.format(formatter);
 	}
 	
+	public static String processResponse(Response response) {
+	    if (response.getEntity() != null) {
+	        try {
+	            InputStream salida = (InputStream) response.getEntity();
```
Comment (ptanhaxon) [inline hds_idp/src/us/hds/ap/idp/service/IDPService.java:89]: What does `salida` mean?

File: hds_idp/src/us/hds/ap/idp/service/IDPService.java (hunk 10 above)
Comment (HoDacVu90) [reply hds_idp/src/us/hds/ap/idp/service/IDPService.java:33]: this function is used for custom rest API to call Training Invoice extraction

File: hds_idp/processes/InvoiceExtraction.p.json (hunk 9 above)
Comment (HoDacVu90) [reply hds_idp/processes/InvoiceExtraction.p.json:254]: It's the UUID of Ivy Rest client which has been called in this process

**Pull request merged**